eitprocessing/filters/butterworth_filters.py

```diff
@@ -0,0 +1,187 @@
+from typing import Literal
+import numpy as np
+import numpy.typing as npt
+from scipy import signal
+from . import TimeDomainFilter
+
+
+MIN_ORDER = 1
+MAX_ORDER = 10
+FILTER_TYPES = ["lowpass", "highpass", "bandpass", "bandstop"]
+
+
+class ButterworthFilter(TimeDomainFilter):
+    """Butterworth filter for filtering in the time domain.
+
+    Generates a low-pass, high-pass, band-pass or band-stop digital Butterworth filter of order
+    `order`.
+
+    ``ButterworthFilter`` is a wrapper of the `scipy.butter()` and `scipy.filtfilt()` functions.
+
+    Args:
+        filter_type: The type of filter to create.
+        cutoff_frequency: Single frequency (lowpass or highpass filter) or tuple containing two
+            frequencies (bandpass and bandstop filters).
+        order: Filter order.
+        sample_frequency: Sample frequency of the data to be filtered.
+        override_order: Whether to raise an exception if the order is larger than the maximum of
+            10. Defaults to False.
+
+    Examples:
+        >>> t = np.arange(0, 100, 0.1)
+        >>> signal = np.sin(t) + 0.1 * np.sin(10 * t)
+        >>> lowpass_filter = ButterworthFilter('lowpass', 45, 4, 250)
+        >>> filtered_signal = lowpass_filter.apply_filter(signal)
+    """
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        filter_type: Literal["lowpass", "highpass", "bandpass", "bandstop"],
+        cutoff_frequency: float | tuple[float, float],
+        order: int,
+        sample_frequency: float,
+        override_order: bool = False,
+    ):
+        if filter_type not in FILTER_TYPES:
+            raise ValueError(
+                "The filter type should be one of "
+                f"{', '.join(FILTER_TYPES)}, not '{filter_type}'."
+            )
+        self.filter_type = filter_type
+
+        if filter_type in ("lowpass", "highpass"):
+            if not isinstance(cutoff_frequency, (int, float)):
+                raise TypeError("`cutoff_frequency` should be an integer or float")
+
+        elif filter_type in ("bandpass", "bandstop"):
+            if not isinstance(cutoff_frequency, tuple):
+                raise TypeError("`cutoff_frequency` should be a tuple")
```

Comment:
Why so strict about not allowing e.g. a `list` instead of a `tuple`? Maybe we can make the check in the module just see if it has `len` of 2 and each value is numeric. I think if those two criteria pass, then there's no reason not to accept. (Potential additional criteria could be that they are non-indentical and/or the first is smaller than the second, but I'm actually not positive that needs to be the case.)

Reply:
The strictness was due to laziness. Now, any sequence of length 2that can be casted to a tuple, except a string, is allowed. This includes lists, arrays, sets, etc, but not `"56"` (which would be casted to `("5", "6")` and might give unexpected behaviour).